## 배열

### 배열선언

#### 일반적 배열 선언

In [2]:
arr1 = [0, 0, 0, 0, 0, 0,]
arr1

[0, 0, 0, 0, 0, 0]

In [3]:
arr2 = [0] * 6
arr2

[0, 0, 0, 0, 0, 0]

#### 리스트 생성자 사용

In [7]:
arr3 = list(range(6))
arr3

[0, 1, 2, 3, 4, 5]

#### 리스트 컨프리헨션 

In [5]:
arr4 = [x for x in range(6)]
arr4

[0, 1, 2, 3, 4, 5]

In [11]:
# 어짜피 사용하지 않는 변수라면 _(사용하지않는 변수)로 변경하라
arr5 = [0 for _ in range(6)]
arr5

[0, 0, 0, 0, 0, 0]

In [14]:
arr4[0] + arr4[1]  + arr4[2]  + arr4[3]  + arr4[4]  + arr4[5] 

15

### 2차원 배열

In [21]:
# 선언
arr6 = [[1,2,3,4], [5,6,7,8], [9,10,11,12]]   # 전체 3x4 배열
arr6

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]

In [17]:
arr6[0]  # [1,2,3,4] 0번 행 값들 전부

[1, 2, 3, 4]

In [20]:
arr6[0][3] # 0행 3열 값을 표시하라

4

In [22]:
arr6[2][3] = 19

In [23]:
arr6

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 19]]

#### 리스트 컨프레헨션

In [31]:
arr7 = [[x for x in range(1 + y * 4, 5 + y * 4)] for y in range(3)]
arr7

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]

In [32]:
[x for x in range(1, 5)]

[1, 2, 3, 4]

### 배열 시간복잡도

- 배열 요소에 접근하는 시간복잡도 : $ O (1) $ . 소요시간이 전혀 걸리지 않음
- 배열에 맨 처음이나 중간에 삽입하는 시간복잡도 : $ O (n) $
- 배열에 맨 처음이나 중간에 삭제하는 시간복잡도 : $ O (n) $

#### 배열 선택시 고려점

- 할당할 메모리 크기 확인. 1차원 배열 최대 1000만개, 2차원 배열 최대 3000 * 3000 (900만개)
- 중간에 데이터 삽입이 많으면 시간복잡도가 높아짐

### 자주 사용하는 리스트 문법

#### append() 사용

In [34]:
my_list = [1,2,3]
my_list.append(4)

In [37]:
my_list

[1, 2, 3, 4]

#### + 연산자 사용

In [38]:
my_list = my_list + [5,6]
my_list

[1, 2, 3, 4, 5, 6]

#### insert() 함수 사용

In [39]:
my_list.insert(3, 999)
my_list

[1, 2, 3, 999, 4, 5, 6]

#### pop() 함수로 데이터 추출

In [40]:
my_list.pop()

6

In [41]:
my_list

[1, 2, 3, 999, 4, 5]

In [42]:
my_list.pop(3)

999

In [44]:
my_list

[1, 2, 3, 4, 5]

#### remove() 로 데이터 삭제. 값으로 삭제

In [47]:
my_list.remove(3) # 리스트에서 3을 지워

ValueError: list.remove(x): x not in list

In [46]:
my_list

[1, 2, 4, 5]

#### 많이 사용하는 문법

In [49]:
fruits = ['apple', 'banana', 'cherry', 'apple', 'orange', 'banana', 'kiwi']

In [50]:
len(fruits)

7

In [51]:
fruits.index('banana')

1

In [52]:
fruits.index('banana', 2)

5

In [53]:
fruits.index('banana', 6)

ValueError: 'banana' is not in list

In [54]:
fruits.sort()

In [55]:
fruits

['apple', 'apple', 'banana', 'banana', 'cherry', 'kiwi', 'orange']

In [56]:
fruits.sort(reverse=True)

In [57]:
fruits

['orange', 'kiwi', 'cherry', 'banana', 'banana', 'apple', 'apple']

In [60]:
fruits.count('mango')

0

### 몸풀기 문제

#### 배열 정렬하기

|입력|출력|
|:--|:--|
|[1,-5,2,4,3]|[-5,1,2,3,4]|
|[2,1,1,3,2,5,4]|[1, 1, 2, 2, 3, 4, 5]|
|[6,1,7]|[1, 6, 7] |

In [ ]:
def solution(arr):
    answer = []
    return answer

In [65]:
# 기본 정렬
def solution(arr):
    arr.sort()
    return arr

In [68]:
#arr1 = [1,-5,2,4,3]
arr1 = [2,1,1,3,2,5,4]
# arr1 = [6,1,7]

solution(arr1)

[1, 1, 2, 2, 3, 4, 5]

In [66]:
# 원본은 보존하면서 작업
def solution(arr):
    answer = sorted(arr, reverse=False)
    return answer

#### sort() 함수 사용않고 정렬알고리즘 구현
- $ O (n^2) $ 시간복잡도

In [85]:
import time

def bubble_sort(arr):   # 버블소트
    n = len(arr)

    for i in range(n):
        for j in range(n - i - 1):  # 한번돌고나면 맨처음 요소는 뛰어넘고 정렬 계속
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]  # 파이썬에서 temp를 만들지 않고 swap가능한 방법

    return arr

def do_sort(arr):  # sort() 사용
    arr.sort()
    return arr

def measure_time(func, arr):   # 시간측정 함수
    start_time = time.time()    # 시작시간 저장
    result = func(arr)  # 함수실행
    end_time = time.time()      # 함수 종료시간 저장

    return (end_time - start_time, result)

arr = list(range(10000))

In [80]:
arr.sort(reverse=True)

##### 버블소트 시간측정

In [86]:
(bubble_time, bubble_result) = measure_time(bubble_sort, arr)

In [83]:
len(bubble_result)

10000

In [84]:
f'버블소트 실행 시간 : {bubble_time:.10f}'

'버블소트 실행 시간 : 3.6535511017'

- 1~9999까지 정렬된 상태에서 버블소트 실행 1.6초 정도 소요
- 내림차순으로 역정렬된 상태에서 버블소트 실행 3.65초 정도 소요

##### 내장 소트 시간측정

In [91]:
arr = list(range(10000))
arr.sort(reverse=True)

In [93]:
sort_time, sort_result = measure_time(do_sort, arr)

In [100]:
f'내장소트 실행 시간 : {sort_time:.70f}'

'내장소트 실행 시간 : 0.0000000000000000000000000000000000000000000000000000000000000000000000'

### 모의 테스트

- 공부할때 사용하는 방법 VS Code. 실제 코테 환경에서는 불가능. 프로그래머스에서 바로 구현(디버깅 불가, 코드어시스턴스 사용불가)

#### 두 개 뽑아서 더하기
- https://school.programmers.co.kr/learn/courses/30/lessons/68644

In [101]:
data = [5,0,2,7]

In [ ]:
def solution(numbers):
    answer = []   # 빈 배열이 미리 생성
    # 두 수를 선택해서 모든 경우의 수를 반복문으로 구함    
    for i in range(len(numbers)):
        for j in range(i+1, len(numbers)):
            # i인덱스와 j인덱스의 값을 더해서 answer에 저장
            answer.append(numbers[i] + numbers[j])
    
    # 중복값 제거
    answer = sorted(set(answer))

    return answer

In [114]:
solution([2,1,3,4,1])

[2, 3, 4, 5, 6, 7]

#### 방문 길이

- https://school.programmers.co.kr/learn/courses/30/lessons/49994

In [ ]:
# 'ULURRDLLU'
# 'LULLLLLLU'

In [137]:
# 좌표가 맵을 벗어나지 않았는지 확인
def is_okay_in_map(nx, ny): 
    # Python 1.0 부터의 기능 
    # chaining 문법 0 <= x < 11
    return 0 <= nx < 11 and 0 <= ny < 11  # 0 ~ 10 사이의 값이 아니면 False

# dir : U, D, L, R 
# 좌표이동 함수
def move_position(dir, x, y):
    if dir.upper() == 'U': # UP
        nx, ny = x, y + 1
    elif dir.upper() == 'D': # DOWN
        nx, ny = x, y - 1
    elif dir.upper() == 'L': # LEFT
        nx, ny = x - 1, y
    elif dir.upper() == 'R': # RIGHT
        nx, ny = x + 1, y

    return nx, ny

def solution(dirs):
    x, y = 5, 5  # 시작위치 5, 5
    answer = set()    # 중복좌표를 제거위해

    for dir in dirs:  # LULLLLLLU 
        nx, ny = move_position(dir, x, y)   # nx, ny 바뀔 좌표
        if is_okay_in_map(nx, ny): # 좌표값이 맵안에 있으며
            answer.add((x, y, nx, ny))  # x,y에서 nx,ny로 이동한것과
            answer.add((nx, ny, x, y))  # nx,ny에서 x,y로 이동한것은 같은 경로
            x, y = nx, ny   # 변경된 좌표로 업데이트
            #    print(answer) 
        else:
            continue

    return int(len(answer) / 2)

In [138]:
solution('LULLLLLLU')

7